apps/prairielearn/src/tests/sharedQuestionPreview.test.ts

```diff
@@ -0,0 +1,104 @@
+import {
+  testQuestionPreviews,
+  testFileDownloads,
+  testElementClientFiles,
+} from './helperQuestionPreview';
+import { config } from '../lib/config';
+import { z } from 'zod';
+import { features } from '../lib/features/index';
+import * as sqldb from '@prairielearn/postgres';
+import * as helperServer from './helperServer';
+import * as syncUtil from './sync/util';
+import { selectCourseById } from '../models/course';
+
+const sql = sqldb.loadSqlEquiv(__filename);
+
+const siteUrl = 'http://localhost:' + config.serverPort;
+const baseUrl = siteUrl + '/pl';
+
+const addNumbers = {
+  id: '',
+  qid: 'addNumbers',
+  type: 'Freeform',
+  title: 'Add two numbers',
+};
+const addVectors = {
+  id: '',
+  qid: 'addVectors',
+  type: 'Calculation',
+  title: 'Addition of vectors in Cartesian coordinates',
+};
+const downloadFile = {
+  id: '',
+  qid: 'downloadFile',
+  type: 'Freeform',
+  title: 'File download example question',
+};
+const customElement = {
+  id: '',
+  qid: 'customElement',
+  type: 'Freeform',
+  title: 'Demo: Custom element',
+};
+const testQuestions = [addNumbers, addVectors, downloadFile, customElement];
+
+describe('Shared Question Preview', function () {
+  this.timeout(60000);
+
+  before('set up testing server', helperServer.before());
+  after('shut down testing server', helperServer.after);
+
+  before('ensure course has question sharing enabled', async () => {
+    await features.enable('question-sharing');
+  });
+
+  before('Get question IDs from database', async function () {
+    for (const testQuestion of testQuestions) {
+      testQuestion.id = await sqldb.queryRow(
+        sql.select_question_id,
+        {
+          qid: testQuestion.qid,
+        },
+        z.string(),
+      );
+      await sqldb.queryAsync(sql.update_shared_publicly, { question_id: testQuestion.id });
+    }
+  });
+
+  describe('Test Public Question Previews', function () {
+    const previewPageInfo = {
+      siteUrl,
+      baseUrl,
+      questionBaseUrl: baseUrl + '/public/course/1/question',
+      questionPreviewTabUrl: '/preview',
+      isStudentPage: false,
+    };
+
+    testQuestionPreviews(previewPageInfo, addNumbers, addVectors);
+
+    testFileDownloads(previewPageInfo, downloadFile, false);
+
+    // TODO: implement
+    // testElementClientFiles(previewPageInfo, customElement);
+  });
+
+  describe('Test Shared Question Previews Within a Course', async function () {
+    const consumingCourseData = syncUtil.getCourseData();
+    consumingCourseData.course.name = 'CONSUMING 101';
+    const consumingCourseResults = await syncUtil.writeAndSyncCourseData(consumingCourseData);
```

Comment:
for some reason, calling `writeAndSyncCourseData` here is giving the following error: 

```
/PrairieLearn/packages/postgres/src/pool.ts:236
      throw new Error('Connection pool is not open');
            ^


Error: Connection pool is not open
    at PostgresPool.getClientAsync (/PrairieLearn/packages/postgres/src/pool.ts:236:13)
    at PostgresPool.queryAsync (/PrairieLearn/packages/postgres/src/pool.ts:512:31)
    at PostgresPool.queryOneRowAsync (/PrairieLearn/packages/postgres/src/pool.ts:535:31)
    at PostgresPool.queryRow (/PrairieLearn/packages/postgres/src/pool.ts:901:32)
    at selectOrInsertCourseByPath (/PrairieLearn/apps/prairielearn/src/models/course.ts:130:16)
    at Object.syncOrCreateDiskToSql (/PrairieLearn/apps/prairielearn/src/sync/syncFromDisk.ts:131:24)
    at syncCourseData (/PrairieLearn/apps/prairielearn/src/tests/sync/util.ts:448:29)
    at Object.writeAndSyncCourseData (/PrairieLearn/apps/prairielearn/src/tests/sync/util.ts:475:29)
    at Suite.<anonymous> (/PrairieLearn/apps/prairielearn/src/tests/sharedQuestionPreview.test.ts:88:36)
```
anyone know why?

Reply:
ok this is very weird, its only happening locally? the CI seems to have passed fine

Reply:
I have no explanation for why this would work in CI. As for why this fails, you might have a bad mental model of how Mocha executes tests. It will first execute all `describe` blocks to discover tests, before/after hooks, etc, and only _then_ will it actually execute the hooks/tests as needed. So, the reason this fails is because your code here runs before the callback in `before('set up testing server', helperServer.before());` is ever executed. As a rule, `describe` callbacks should not have side effects and should not be `async` (as Mocha won't await them).

To fix this, you can move all of this to a `before` hook here, which will execute after the `before` hook on the parent `describe` suite.

Reply:
A thought about why this works in CI: are you running this locally with something like `yarn mocha src/tests/sharedQuestionPreview.test.ts`? This actually uses a subtly different execution model (it doesn't run multiple processes for parallel execution), which I think explains the difference. You can reproduce passing tests locally with `make test-prairielearn`, which is what is run in CI. In fact, if you grep through the output, you'll see that `Test Shared Question Previews Within a Course` doesn't appear in the output, meaning that these tests were never even run.

Reply:
Thanks Nathan, you were right about everything :) 

I believe its all fixed now.